Modules/_csv.c

```diff
@@ -238,21 +238,29 @@ _set_char(const char *name, Py_UCS4 *target, PyObject *src, Py_UCS4 dflt)
         if (src != Py_None) {
             Py_ssize_t len;
             if (!PyUnicode_Check(src)) {
-                PyErr_Format(PyExc_TypeError,
-                    "\"%s\" must be string, not %.200s", name,
-                    Py_TYPE(src)->tp_name);
+                if (strcmp(name, "delimiter") == 0) {
+                    PyErr_Format(PyExc_TypeError,
+                        "\"%s\" must be string, not %.200s", name,
+                        Py_TYPE(src)->tp_name);
+                }
+                else {
+                    PyErr_Format(PyExc_TypeError,
+                        "\"%s\" must be string or None, not %.200s", name,
+                        Py_TYPE(src)->tp_name);
+                }
                 return -1;
             }
             len = PyUnicode_GetLength(src);
-            if (len > 1) {
+            if (len != 1) {
                 PyErr_Format(PyExc_TypeError,
                     "\"%s\" must be a 1-character string",
                     name);
                 return -1;
             }
             /* PyUnicode_READY() is called in PyUnicode_GetLength() */
-            if (len > 0)
+            else {
                 *target = PyUnicode_READ_CHAR(src, 0);
+            }
```

Comment:
I'm wondering if it's ok to backport this. What happens currently if the delimiter is of length 0? Are there situations where the constructor ignores it and then it doesn't matter because the delimiter is not used?

Reply:
@iritkatriel 

Due to the following code lines.

https://github.com/python/cpython/blob/a25dcaefb7c4eb0767a112cd31fe0b055f168844/Modules/_csv.c#L254-L255

https://github.com/python/cpython/blob/a25dcaefb7c4eb0767a112cd31fe0b055f168844/Modules/_csv.c#L458-L461

It is failed to be constructed.

Reply:
I don't follow.  This doesn't fail:

```
>>> import csv
>>> class mydialect(csv.Dialect):
...              delimiter = ";"
...              escapechar = '\\'
...              doublequote = False
...              skipinitialspace = True
...              lineterminator = '\r\n'
...              quoting = csv.QUOTE_NONE
... 
>>> mydialect.escapechar = ""
>>> mydialect()
<__main__.mydialect object at 0x1017f3a20>
>>> 

```

Reply:
I think from the commit log that @serhiy-storchaka might be familiar with this feature.

Reply:
It seems that `escapechar=""` and `quotechar=""` have the same effect as `escapechar=None` and `quotechar=None` (and also `escapechar="\0"` and `quotechar="\0"`). I do not know whether it is intentional or an artifact of implementation. Ask @smontanaro.

Reply:
> What happens currently if the delimiter is of length 0?

@iritkatriel 
I thought that we talk about `delimiter` not `escapechar`.
For the `delimiter` it will fail as I said, but `escapechar` it does not fail.

`````

>>> mydialect.delimiter = ""
>>> mydialect()
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py", line 50, in _validate
    _Dialect(self)
TypeError: "delimiter" must be a 1-character string

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py", line 46, in __init__
    self._validate()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py", line 53, in _validate
    raise Error(str(e))
_csv.Error: "delimiter" must be a 1-character string
`````

Reply:
Anyway as @serhiy-storchaka suggested, I will request the review to @smontanaro